.github/workflows/build_and_test.yml

```diff
@@ -18,58 +20,86 @@ jobs:
-        if: always()
-      - name: Compare out_ir.py
-        run: diff -u out_ir.py python/experiments/out_ir.py
-        if: always()
-
+        run: dist/kotlinc/bin/kotlinc-py -libraries dist/kotlinc/lib/kotlin-stdlib-js.jar -Xir-produce-js -output python/experiments/out_ir.py python/experiments/python.kt
+        if: ${{ matrix.testTask == 'pythonTest' }}
       - name: Compile python.kt to JavaScript
-        run: dist/kotlinc/bin/kotlinc-js -libraries dist/kotlinc/lib/kotlin-stdlib-js.jar -Xir-produce-js -Xir-property-lazy-initialization -output out-ir.js python/experiments/python.kt
-        if: always()
-      - name: Compare out-ir.js
-        run: diff -u out-ir.js python/experiments/out-ir.js
-        if: always()
+        run: dist/kotlinc/bin/kotlinc-js -libraries dist/kotlinc/lib/kotlin-stdlib-js.jar -Xir-produce-js -Xir-property-lazy-initialization -output python/experiments/out-ir.js python/experiments/python.kt
+        if: ${{ matrix.testTask == 'pythonTest' }}
 
       - name: Generate stats about missing IR mapping
-        run: less python/experiments/out_ir.py | grep -Po "visit[a-zA-Z0-9_]+" | sort | uniq -c | sort -nr > missing.txt
-        if: always()
-      - name: Compare missing.txt
-        run: diff -u missing.txt python/experiments/missing.txt
-        if: always()
+        run: less python/experiments/out_ir.py | grep -Po "visit[a-zA-Z0-9_]+" | sort | uniq -c | sort -nr > python/experiments/missing.txt
+        if: ${{ matrix.testTask == 'pythonTest' }}
 
       - name: Run end-to-end tests
         run: JDK_9="$JAVA_HOME" python/e2e-tests/run.sh
-        if: always()
+        if: ${{ matrix.testTask == 'pythonTest' }}  # TODO: run e2e tests for MicroPython too (#85)
 
       - name: Run box tests (succeed even if they fail)
         run: JDK_9="$JAVA_HOME" ./gradlew :python:box.tests:${{ matrix.testTask }} || true
-        if: always()
       - name: Generate box tests reports
-        run: python/experiments/generate-box-tests-reports.main.kts --test-task=${{ matrix.testTask }} --failed-tests-report-path=failed-tests.txt --box-tests-report-path=box-tests-report.tsv --failure-count-report-path=failure-count.tsv --git-history-plot-path=git-history-plot.svg
-        if: always()
-      - name: Compare failed-tests.txt
-        run: diff -u failed-tests.txt python/box.tests/reports/${{ matrix.testTask }}/failed-tests.txt
-        if: always()
-      - name: Compare box-tests-report.tsv
-        run: diff -u box-tests-report.tsv python/box.tests/reports/${{ matrix.testTask }}/box-tests-report.tsv
-        if: always()
-      - name: Compare failure-count.tsv
-        run: diff -u failure-count.tsv python/box.tests/reports/${{ matrix.testTask }}/failure-count.tsv
-        if: always()
+        run: python/experiments/generate-box-tests-reports.main.kts --test-task=${{ matrix.testTask }}
```

Comment:
I anticipate two risks that follows changing the approach from "compare" to "generate".

1. if our reports include some random piece (like we already had: memory address, number of failed allocated bytes, and so on), each change will now trigger the automated commit that will add changes just in the random pieces.
2. So far if our local machines behaved in a certain way, and GH Actions workers in a different way, the "compare" approach allowed us to catch it if it resulted in different outputs. After this change, GH Actions environment will be distinguished in some way, if you know what I mean.

I'm still for going with this change, but it's something to keep in mind. Usually a proper PR review should suffice, we should be able to spot potential random places.

Reply:
OK, I understand your concerns. Do you have any ideas how we can detect such non-deterministic stuff automatically? And should we actually?

Maybe it's not a big problem since our current goal is to make box tests pass. We can, for example, use runners with other two OSes and compare outputs. However, maybe we shouldn't bother now.

Reply:
We are doing something similar to what you're talking about in refreshVersions:

The CI runs some tests, and if the content of some files differ with what it computed, it fails.

Running the same tests locally either fail, but leave the expected files in the file system with another name, so you cam copy the content into the actual files where the tests run the comparison, or, they directly update the files. Both ways, you can commit the changes, and the CI will ensure this changes are correct.

Now, for your case, if you don't want to have to always run the tests locally, you could add a manually triggered GitHub Action (called workflow_dispatch trigger or something), which would run the git commands to run the tests with the same mode as when you run it locally.

If you search in the refreshVersions repo, you'll find the `isInCi` function, which enables us to have that slightly different behavior for the tests.

Reply:
Wow, I didn't think we can apply workflow_dispatch here. We could avoid editing the current build_and_test action, but create one more action like update_generated_files which can be invoked by a button that will update the generated files.

@krzema12, wdyt?

Reply:
What did you think workflow_dispatch was limited to?

Reply:
Nothing. I just wasn't creative enough to understand it can be useful for us. Now I see it as a good solution.

Thanks for mentioning it and giving this idea, btw ;)

Reply:
To sum up, we have two options:
1. Trigger the regeneration of test result files automatically, in each CI run - implemented now in this PR.
2. Trigger regeneration manually, on demand.

I don't have a strong opinion here, but if you have @SerVB and option 2 would be more comfortable to you, then please go ahead :) If I were to decide, I'd stay with what we have in this PR now (eliminates a need to remember about running something extra) and see how it works, to not spend too much more time on it. If it turns out it's better to make it the on-demand job, we can do it later.

Reply:
It's also important for me to not go too deep with improvements and not overengineer, so that we can have a healthy balance between quality and delivering features.

Reply:
> It's also important for me to not go too deep with improvements and not overengineer, so that we can have a healthy balance between quality and delivering features.

Well, since it's a pet project, I would like to try some new stuff out: this is fun and can be handy. Moreover, I don't think that making CI closer to ideal is overengineering, it least in this case...

I hope you won't mind much :)

> eliminates a need to remember about running something extra

The failed CI will be the reminder ;)

> Trigger regeneration manually, on demand.

I've just done this way, testing. Will update the PR later and request your review.